### DISCUSSION TITLE
Alignment issue using EigenPy for custom type

### DISCUSSION THREAD

**Discussion by ofloveandhate** [stack-of-tasks/eigenpy#387] [Q&A]
Hi!  I use EigenPy to expose Eigen with my custom numeric type (Boost::Multiprecision complexes, which i'll call `mpc`) with dynamically sized matrices into Python.  

I'm encountering an issue in Python where I have a numpy array of `mpc`, as output from one function call, and I'm feeding it into another function expecting an Eigen matrix of `mpc`.  But I'm getting 

```
Assertion failed: (is_aligned(raw_ptr, EIGENPY_DEFAULT_ALIGN_BYTES) && "The pointer is not aligned."), function allocate, file eigen-allocator.hpp, line 229.
Abort trap: 6
```

I'm looking for advice on how to go about debugging this.  I can produce backtraces, etc, but I figured I'd start by asking for advice since I suspect you've encountered this before -- I bet you have the assert there to help find this issue, and maybe there's a simple "thing I should be doing".

I'd appreciate any help you can give me.  Thanks a billion!

**Comment by ofloveandhate** [comment]
Is this related to Issue #330 ?

**Reply by ofloveandhate** [reply]
moving from 3.0.0 to 3.1.1 did not change this behaviour.  (i posted an earlier comment suggesting it might, then deleted it because I was wrong)

**Reply by ofloveandhate** [reply]
Here's a backtrace on the `storage` object at the point i'm hitting the assert:

```
(lldb) p *storage
(boost::python::converter::rvalue_from_python_storage<Eigen::Matrix<boost::multiprecision::number<boost::multiprecision::backends::mpc_complex_backend<0> >, -1, 1> >) {
  stage1 = {
    convertible = 0x000000012772b510
    construct = 0x0000000103566338 (_pybertini.so`eigenpy::eigen_from_py_impl<Eigen::Matrix<boost::multiprecision::number<boost::multiprecision::backends::mpc_complex_backend<0u>, (boost::multiprecision::expression_template_option)1>, -1, 1, 0, -1, 1>, Eigen::MatrixBase<Eigen::Matrix<boost::multiprecision::number<boost::multiprecision::backends::mpc_complex_backend<0u>, (boost::multiprecision::expression_template_option)1>, -1, 1, 0, -1, 1>>>::construct(_object*, boost::python::converter::rvalue_from_python_stage1_data*) at eigen-from-python.hpp:409:3)
  }
  storage = {
    data = {
      data_ = (buf = "\xbc\U0000001d\xfc\U00000003\U00000001\0\0\0 \x96\xc0\0\0`", align_ = boost::type_with_alignment<16>::type @ 0x0000600001338df0)
    }
    bytes = "\xbc\U0000001d\xfc\U00000003\U00000001\0\0\0 \x96\xc0\0\0`"
  }
}

```

**Reply by ofloveandhate** [reply]
i suspect i'm running into a dynamic vs fixed size alignment thing?

**Reply by ofloveandhate** [reply]
Well, I was able to work with the debugger, and indeed the pointer is not 16-aligned.  So now I need to go back in my code and find out where that container was made, and make sure it's 16-alignining at that location.  

I'm so confused, though, because the Eigen documentation suggests repeatedly that fixed-size containers exhibit alignment issues.  But my code only works with dynamically sized containers, and indeed this problem I'm encountering is with dynamic-length vectors, not fixed-size.  Soooo...  I move forward trying to chase down the source of this vector and make sure its creator knows to 16-align them.

**Reply by jcarpent** [reply]
Could you provide a reproducible example?

**Reply by ofloveandhate** [reply]
Thanks for your reply.  I will increase the priority of this in my list of todo's.  A MWE would indeed be great.  

As I work toward one, have you experienced alignment issues ever between arrays made with numpy and functions in EigenPy / Pinnochio / etc?

**Reply by ofloveandhate** [reply]
I'm trying to write a reproducible example, but I'm struggling to get eigenpy / my previous example to compile on my machine, due to an issue with finding Python.  argh.

**Reply by ofloveandhate** [reply]
I finally was able to get my example "library" compiling and importing correctly using `conda`, so now I can make progress on my MWE for my alignment issue.
